test/MyFactory.js

```diff
@@ -0,0 +1,385 @@
+      assert.isOk(newUserBBalance.eq(userBBalance.add(amount)));
+    });
+
+    it('should work for proxy', async () => {
+      const amount = toBN(100);
+      const userBBalance = await myCollectible.balanceOf(userB, PREMIUM);
+      await myFactory.safeTransferFrom(
+        ADDRESS_ZERO,
+        userB,
+        PREMIUM,
+        100,
+        "0x0",
+        { from: proxyForOwner }
+      );
+      const newUserBBalance = await myCollectible.balanceOf(userB, PREMIUM);
+      assert.isOk(newUserBBalance.eq(userBBalance.add(amount)));
+    });
+
+    it('should not be callable by non-owner() and non-proxy', async () => {
+      const amount = toBN(100);
+      await truffleAssert.fails(
+        myFactory.safeTransferFrom(
+          ADDRESS_ZERO,
+          userB,
+          PREMIUM,
+          amount,
+          "0x0",
+          { from: userB }
+        ),
+        truffleAssert.ErrorType.revert,
+        'MyFactory#_mint: CANNOT_MINT_MORE'
+      );
+    });
+  });
+
+  describe('#isApprovedForAll()', () => {
+    it('should approve owner as both _owner and _operator', async () => {
+      assert.isOk(
+        await myFactory.isApprovedForAll(owner, owner)
+      );
+    });
+
+    it('should not approve non-owner as _owner', async () => {
+      assert.isNotOk(
+        await myFactory.isApprovedForAll(userA, owner)
+      );
+      assert.isNotOk(
+        await myFactory.isApprovedForAll(userB, userA)
+      );
+    });
+
+    it('should not approve non-proxy address as _operator', async () => {
+      assert.isNotOk(
+        await myFactory.isApprovedForAll(owner, userB)
+      );
+    });
+
+    it('should approve proxy address as _operator', async () => {
+      assert.isOk(
+        await myFactory.isApprovedForAll(owner, proxyForOwner)
+      );
+    });
+
+    it('should reject proxy as _operator for non-owner _owner', async () => {
+      assert.isNotOk(
+        await myFactory.isApprovedForAll(userA, proxyForOwner)
+      );
+    });
+  });
+
+  /**
+   * NOTE: This check is difficult to test in a development
+   * environment, due to the OwnableDelegateProxy. To get around
+   * this, in order to test this function below, you'll need to:
+   *
+   * 1. go to MyFactory.sol, and
+   * 2. modify _isOwnerOrProxy
+   *
+   * --> Modification is:
+   *      comment out
+   *         return owner() == _address || address(proxyRegistry.proxies(owner())) == _address;
+   *      replace with
+   *         return true;
+   * Then run, you'll get the reentrant error, which passes the test
+   **/
+
+  describe('Re-Entrancy Check', () => {
+    it('Should have the correct factory address set',
+       async () => {
+         assert.equal(await attacker.factoryAddress(), myFactory.address);
+       });
+
+    // With unmodified code, this fails with:
+    //   MyFactory#_mint: CANNOT_MINT_MORE
+    // which is the correct behavior (no reentrancy) for the wrong reason
```

Comment:
nice catch!

Reply:
Thank you. This is something to watch for - it's currently secure but if anything underneath changes then it could expose the attack.